### PR TITLE
Dashboard upgrade

### PR DESCRIPTION
**Pull request by ericmatte**
## Description

This was becoming a little big, so I'm pushing this PR right now, but there is still some work to do.

# New features:
- Cleaner dialog input
- `Y` and `N` on the keyboard trigger yes and no answer
- Simple textbox to enter the player name (and also buttons for Yes/No)
- **Auto-reconnect to ros on connection lost**
- Dual monitor for the kinect.
- **The old dashboard with more options is still available at** `/debug`

# What remains?
- Testing with kinect images or rosbag

# Screenshots:
image

image

## PR Checklist

- [x] Documentation
- [x] Tests  **Partly done

### CODE REVIEW

**Comment by ericmatte** [comment]
For the auto-reconnect, what message do you guys prefer?
1. Connection to ROS lost.
2. The project is currently rebooting. Please wait.
3. Something else

File: src/dashboard/src/devine_dashboard/js/dashboard/canvas_drawer.js [hunk 1]
```diff
@@ -0,0 +1,319 @@
...
+    
+    this.history = this._initHistoryBuffers();
+    this.changeImageSource('image_raw');
+
+    this._initConfidenceReset();
+  }
+
+  /** Draw current image on canvas */
+  draw() {
+    // TODO: const draw = throttle(function draw() {}, 100);
```
Comment (ericmatte) [inline src/dashboard/src/devine_dashboard/js/dashboard/canvas_drawer.js]: @tihbe I've removed that for now, but is it necessary to keep it?

Comment (tihbe) [reply src/dashboard/src/devine_dashboard/js/dashboard/canvas_drawer.js]: Not anymore no, we're throttling openni directly

Comment (ericmatte) [reply src/dashboard/src/devine_dashboard/js/dashboard/canvas_drawer.js]: Perfect

**Comment by tihbe** [comment]
> For the auto-reconnect, what message do you guys prefer?
> 
> 1. Connection to ROS lost.
> 2. The project is currently rebooting. Please wait.
> 3. Something else

Number 1 is fine

File: src/dashboard/src/devine_dashboard/html/demo/demo.html [hunk 2]
```diff
@@ -0,0 +1,27 @@
+{% import 'macros/camera_feed.html' as camera_macro %}
+{% import 'macros/console.html' as console_macro %}
+
+<div class="demo">
+    <div class="modal fade" id="reconnection-modal" tabindex="-1" role="dialog">
+        <div class="modal-dialog" role="document">
+            <div class="modal-content">
+            <div class="modal-header">
+                <h5 class="modal-title">Connection to ROS lost.</h5>
```
Comment (tihbe) [inline src/dashboard/src/devine_dashboard/html/demo/demo.html]: I don't think it's necessary to end with a period in a modal title

File: src/dashboard/src/devine_dashboard/html/demo/demo.html [hunk 3]
```diff
@@ -0,0 +1,27 @@
...
+    <div class="live-feed">
+        {{ camera_macro.camera('Live feed', 'live_feed') }}
+    </div>
+    <div class="segmentation">
+        {{ camera_macro.camera('Objects segmentation', 'segmentation_feed') }}
+    </div>
+    <div class="dialog">
+        {{ console_macro.console('Dialog', false, true) }}
+    </div>
+</div>
```
Comment (tihbe) [inline src/dashboard/src/devine_dashboard/html/demo/demo.html]: newline

File: src/dashboard/src/devine_dashboard/html/dashboard/dashboard.html [hunk 4]
```diff
@@ -1,13 +1,16 @@
...
   <div class="col-sm">
     {% include 'dashboard/publisher.html' %} 
     {% include 'dashboard/subscribers/dialog.html' %} 
     {% include 'dashboard/subscribers/image_dispatcher.html' %} 
   </div>
   <div class="col-sm">
-    {% include 'dashboard/subscribers/kinect.html' %}
+    {{ kinect_macro.camera('Kinect Camera', 'kinect_image', true) }}
   </div>
 </div>
```
Comment (tihbe) [inline src/dashboard/src/devine_dashboard/html/dashboard/dashboard.html]: newline

File: src/dashboard/src/devine_dashboard/css/styles.css [hunk 5]
```diff
@@ -67,3 +53,74 @@ body > div {
...
+  margin-right: 16px;
+  padding: 0 !important;
+  height: 31px;
+}
+.dialog-inputs .input-group-text {
+  background-color: #303030;
+}
+.dialog-inputs button {
+  line-height: 1;
+}
```
Comment (tihbe) [inline src/dashboard/src/devine_dashboard/css/styles.css]: newline

File: src/dashboard/src/devine_dashboard/js/dashboard/canvas_drawer.js [hunk 6]
```diff
@@ -0,0 +1,319 @@
...
+      }
+    }
+  }
+
+  /**
+   * Map body part with one another (e.g.: ear with eye).
+   * Adaptation and refactor of function draw_humans in tf_pose/estimator.py
+   * @param {array} humans - The list of humans detected.
+   *
+   * @see CocoPairs in tf_pose/common.py
```
Comment (tihbe) [inline src/dashboard/src/devine_dashboard/js/dashboard/canvas_drawer.js]: The comment about cocopairs should maybe stay with the `BODY_PARTS` constant. Up to you

File: src/dashboard/src/devine_dashboard/js/dashboard/subscribers/queries_answers.js [hunk 7]
```diff
@@ -0,0 +1,89 @@
...
+      this.topics.ttsAnswer.publish(
+        new ROSLIB.Message({
+          original_query: query,
+          probability: 1.0,
+          text: text
+        })
+      );
+      
+      query.answered = true;
+      query_answered = true;
```
Comment (tihbe) [inline src/dashboard/src/devine_dashboard/js/dashboard/subscribers/queries_answers.js]: I'm not sure but we might need a break here, when the query is answered.

File: src/dashboard/src/devine_dashboard/js/demo/demo.js [hunk 8]
```diff
@@ -0,0 +1,43 @@
...
+  });
+
+  let countdown;
+  /** On connection lost, try to reconnect each 3 seconds. */
+  function tryAutomaticReconnect() {
+    $('#reconnection-modal').modal({ backdrop: 'static'});
+
+    let seconds = 3;
+    countdown = setInterval(() => {
+      if (seconds === 0) {
```
Comment (tihbe) [inline src/dashboard/src/devine_dashboard/js/demo/demo.js]: `seconds <== 0` is safer

File: src/dashboard/src/devine_dashboard/js/dashboard/ros.js [hunk 9]
```diff
@@ -18,21 +18,24 @@ export class RosTopic extends ROSLIB.Topic {
   }
 }
 
+/** Try to reconnecto to ROS */
```
Comment (tihbe) [inline src/dashboard/src/devine_dashboard/js/dashboard/ros.js]: Reconnecto
image

**Comment by tihbe** [comment]
I'll try to test it whenever I can, Amazing job !!

File: src/dashboard/src/devine_dashboard/js/dashboard/canvas_drawer.js (hunk 6 above)
Comment (ericmatte) [reply src/dashboard/src/devine_dashboard/js/dashboard/canvas_drawer.js]: Good point

File: src/dashboard/src/devine_dashboard/js/dashboard/subscribers/queries_answers.js (hunk 7 above)
Comment (ericmatte) [reply src/dashboard/src/devine_dashboard/js/dashboard/subscribers/queries_answers.js]: Ah good catch!! I did sometime get a glitch where one answer answered multiple queries in the dashboard. That must be it.

**Pull request merged**